### PR TITLE
[WASM] update nn_ops.js with tflite optimization

### LINKED ISSUE #331: [WASM] update nn ops
Current nn ops is compiled from source code of Android O. 
It needs to be updated to Android P which shares the ops implementation with TensorFlow Lite.

### PR DESCRIPTION
**Pull request by Wenzhao-Xiang**
Fix #331 

Test:
```
passes: 392   failures: 176    duration: 3.97s =>
                                passes: 392   failures: 176    duration: 3.97s
```

The performance compare is here https://github.com/intel/webml-polyfill/issues/331#issuecomment-442717929. Seem to have a 2%~3% performance improvement with tflite optimization.

And the source code is here [webml-polyfill-wasm](https://github.com/Wenzhao-Xiang/webml-polyfill-wasm).

### CODE REVIEW

**Comment by Wenzhao-Xiang** [comment]
https://github.com/Wenzhao-Xiang/webml-polyfill-wasm/commit/e429f03035c7f98be38fbb7dc4eeeff0a6f56f95 : Mainly solve the compiling issues, and remove the unnecessary dependency.

https://github.com/Wenzhao-Xiang/webml-polyfill-wasm/commit/9bb14b4068e0acdd816a233932916bb9b07257b6: Now the NNAPI for softmax is different from the source, whitch have an extra param "axis". To match this, delete this param and set it inner the function with 
```c
int32_t axis = ndim - 1;
```
(see it in https://github.com/Wenzhao-Xiang/webml-polyfill-wasm/blob/8375e16b22b1ef015958a1be40590597cc80a8ac/ml/nn/common/operations/Activation.cpp#L114)

**Comment by huningxin** [comment]
It's a great work. Thanks for the PR! @Wenzhao-Xiang 

Some suggestions:
1. as we adopt the `tflite::optimized_ops`, can we remove the nn layer? It will lead fewer dependencies.
2. can you merge the wasm build process into the webml-polyfill build? Please refer to onnx.js: https://github.com/Microsoft/onnxjs/blob/master/tools/build.ts
3. can you make tensorflow as git submodule, similar to https://github.com/Microsoft/onnxjs/tree/master/deps

**Comment by Wenzhao-Xiang** [comment]
There are some updates in [webml-polyfill-wasm](https://github.com/Wenzhao-Xiang/webml-polyfill-wasm):
https://github.com/Wenzhao-Xiang/webml-polyfill-wasm/commit/e75a28fd96a76da936bb1b660804e22cff71f5ff : set tensorflow(r1.11) as a submodule and use download_dependencies.sh to download tensorflow dependencies
https://github.com/Wenzhao-Xiang/webml-polyfill-wasm/commit/902bf6570b8d3eb3a25e7efd3b8e02f4d44131ae : update the ops dependencies to match the new submodule

And the building steps can see in [README.md](https://github.com/Wenzhao-Xiang/webml-polyfill-wasm/blob/master/README.md)

**Comment by Wenzhao-Xiang** [comment]
I matched the source code with the newest tensorflow (master) and set tensorflow as a submodule (https://github.com/Wenzhao-Xiang/webml-polyfill-wasm/commit/6c12ebffdba4f82b9500e17b60017fe412640e12  and  https://github.com/Wenzhao-Xiang/webml-polyfill-wasm/commit/78c51f9007ef5670d859ad8eed5ade2beaeb0855) . Now, it's more convenient to update the dependency. 

And also, now we can use the command **"npm run build-wasm"** to build the nn_ops.js in Linux.

**Comment by huningxin** [comment]
Great work! @Wenzhao-Xiang 

I am wondering why not add the source code of https://github.com/Wenzhao-Xiang/webml-polyfill-wasm into this repo? It looks like light weight now.

**Comment by Wenzhao-Xiang** [comment]
@huningxin Thanks for review! 

I will move the repo(https://github.com/Wenzhao-Xiang/webml-polyfill-wasm ) to the directory 
**webml-polyfill/src/nn/wasm/src**, is that ok?

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[submodule "deps/emsdk"]
+	path = deps/emsdk
+	url = https://github.com/juj/emsdk.git
+[submodule "src/nn/wasm/src/external/tensorflow"]
+	path = src/nn/wasm/src/external/tensorflow
+	url = https://github.com/tensorflow/tensorflow.git
```
Comment (huningxin) [inline .gitmodules]: I noticed that the commit of tensorflow you added is https://github.com/tensorflow/tensorflow/tree/04e8759ee2416baac1f31f6a27cb49a8b6051e19, any reasons to add this commit? Do we need to align to the version used by Android NN  or some releases of TensorFlow-Lite?

File: src/nn/wasm/src/common/operations/Conv2D.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    ANDROID_NN_CONV_PARAMETERS(float)
+
+    float output_activation_min, output_activation_max;
+    CalculateActivationRangeFloat(activation, &output_activation_min,
+                                  &output_activation_max);
+    
+    int32_t dilationWidthFactor = 1, dilationHeightFactor = 1;
+
+    tflite::optimized_ops::Conv(
```
Comment (huningxin) [inline src/nn/wasm/src/common/operations/Conv2D.cpp]: Can we expose the `tflite::optimized_ops::Conv` and `tflite::Dims` directly? So the js code can call it without this Android NN wrapper code.

Comment (huningxin) [reply src/nn/wasm/src/common/operations/Conv2D.cpp]: If it works, we can apply it to other ops and simplify our binding code.

**Comment by huningxin** [comment]
> Done! Please help take a look!

good work! I've reviewed the PR and think there might be still room to improve. Please take a look at my comments.

File: .gitmodules (hunk 1 above)
Comment (Wenzhao-Xiang) [reply .gitmodules]: No, we don't need to align the version. I used 
```
git submodule add https://github.com/tensorflow/tensorflow.git src/nn/wasm/src/external/tensorflow
```
And seems it fetch the newest version of tensorflow automatically.
Do I have a wrong use of `git submodule add`?

File: src/nn/wasm/src/common/operations/Conv2D.cpp (hunk 2 above)
Comment (pinzhenx) [reply src/nn/wasm/src/common/operations/Conv2D.cpp]: I don't know if I understand it correctly. Do you mean calling TFLite code directly in the binding code?

If that's the case, we may still need to move these NN code to the bindings in dealing these parameters. Or should we make these Android wrappers`inline`?

Comment (huningxin) [reply src/nn/wasm/src/common/operations/Conv2D.cpp]: No. I mean call tflite functions directly in JS. For example, call `tflite::optimized_ops::Conv` 
 [here](https://github.com/intel/webml-polyfill/blob/master/src/nn/wasm/PreparedModel.js#L196). No need NN wrapper code anymore.

Comment (pinzhenx) [reply src/nn/wasm/src/common/operations/Conv2D.cpp]: Ok, I see what you mean. We'll keep working on it.

Comment (Wenzhao-Xiang) [reply src/nn/wasm/src/common/operations/Conv2D.cpp]: Thanks, I will first have a try with Conv2D!

**Comment by Wenzhao-Xiang** [comment]
@huningxin 
And I found tensorflow lite now use `RuntimeShape` as the type of Shape actually, not `Dims<4>` (`Dims<4>` will be converted to `RuntimeShape`, if used) . `RuntimeShape` is a class that is a little bit complex. And I'll also try to expose it later.

**Comment by Wenzhao-Xiang** [comment]
Finished with redesigning interface of Conv2D and  exposing it for WASM backend.

`tflite::RuntimeShape` is exposed as following:
```c
class_<tflite::RuntimeShape>("RuntimeShape")
    .constructor<int>(select_overload<tflite::RuntimeShape(int)>([](int dimensions_count) {
        return tflite::RuntimeShape(dimensions_count);
      }
    ))
    .property("dims", &binding_utils::getRuntimeShapeDimensions, &binding_utils::setRuntimeShapeDimensions)
    .property("size", &binding_utils::getRuntimeShapeSize, &binding_utils::setRuntimeShapeSize)
    ;
```


`tflite::ConvParams` is exposed as following:
```c
class_<tflite::ConvParams>("ConvParams")
    .constructor<>()
    .property("padding_values", &binding_utils::getPaddingValues, &binding_utils::setPaddingValues)
    .property("strides", &binding_utils::getStrides, &binding_utils::setStrides)
    .property("dilation_factors", &binding_utils::getDilationFactor, &binding_utils::setDilationFactor)
    .property("float_activation_range", &binding_utils::getActivationRange, &binding_utils::setActivationRange)
    ;
```
And the definition is:
**padding_values = [paddingWidth, paddingHeight]
strides = [strideWidth, strideHeight]
dilation_factors = [dilation_width_factor, dilation_height_factor]
float_activation_range = [float_activation_min, float_activation_max]**


And now directly expose `tflite::optimized_ops::Conv()` as following:
```c
bool convFloat32Wrapper(tflite::ConvParams& op_params, 
                          const tflite::RuntimeShape& inputShape, const intptr_t inputData, 
                          const tflite::RuntimeShape& filterShape, const intptr_t filterData, 
                          const tflite::RuntimeShape& biasShape, const intptr_t biasData, 
                          const tflite::RuntimeShape& outputShape, intptr_t outputData,
                          const tflite::RuntimeShape& im2colShape, intptr_t im2colData) {
    tflite::optimized_ops::Conv(
            op_params, 
            inputShape, (const float*)inputData, 
            filterShape, (const float*)filterData, 
            biasShape, (const float*)biasData, 
            outputShape, (float*)outputData, 
            im2colShape, (float*)im2colData);
    return true;
  }
```


And move the params inition to JavaScript code:
```js
let im2colDepth = filterWidth * filterHeight * inDepth;
let im2colDimentions = [outBatch, outHeight, outWidth, im2colDepth];
let im2colValue = new Float32Array(outBatch * outHeight * outWidth * im2colDepth);
let operand = {
    type: OperandCode.TENSOR_FLOAT32,
    dimensions: im2colDimentions,
    numberOfConsumers: 0,
    lifetime: OperandLifetime.CONSTANT_REFERENCE,
    value: im2colValue
    }

// init convParams
let convParams = new nn_ops.ConvParams;
convParams.padding_values = [paddingLeft, paddingTop];
convParams.strides = [strideWidth, strideHeight];
onvParams.dilation_factors = [1, 1];
let float_activation_min = calculateActivationRangeFloat(activation).activation_min;
let float_activation_max = calculateActivationRangeFloat(activation).activation_max;
convParams.float_activation_range = [float_activation_min, float_activation_max];
let im2colShape = this._allocateDims(operand).RuntimeShape;
let im2colData = this._allocateTensor(operand);
```

Since our API allocate all the tensor shape at first, the `convPrepare()` just play a role in **error check**.Therefore, plan to remove it and move the error check to JS code.

**Comment by Wenzhao-Xiang** [comment]
Fix #386 with the new WASM backend:

Test:  (don't include `End2End Test/ONNX Models`)
```
passes: 413  failures: 193  duration: 5.99s =>
                                passes: 414  failures: 192  duration: 6.66s
```

**Comment by Wenzhao-Xiang** [comment]
Done! 
After remove all the extra check and prepare part, the new nn_ops.js is now just **169k**, reduced about **254k** (reduced about **60%**)  from original **423k**.
@huningxin  Please help take a look!

File: src/nn/wasm/src/bind/src/binding.cpp [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+  // Operation wrappers.
+  bool addFloat32Wrapper(const tflite::ArithmeticParams& op_params,
+                                  const tflite::RuntimeShape& input1_shape, const intptr_t input1_data, 
+                                  const tflite::RuntimeShape& input2_shape, const intptr_t input2_data, 
+                                  const tflite::RuntimeShape& output_shape, intptr_t output_data) {
+    tflite::optimized_ops::Add(op_params,
+                               input1_shape, (const float*) input1_data,
+                               input2_shape, (const float*) input2_data,
+                               output_shape, (float*) output_data);
+    return true;
```
[the remark this answers is not in the document]
Comment (Wenzhao-Xiang) [reply src/nn/wasm/src/bind/src/binding.cpp]: This return is used for JS code to throw errors:
```js
if (!success) {
          throw new Error(`${funcName} fails`);
        }
```
Or you mean we don't need throw errors there?

File: src/nn/wasm/src/bind/src/binding.cpp [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+  class_<tflite::RuntimeShape>("RuntimeShape")
+    .constructor<int>(select_overload<tflite::RuntimeShape(int)>([](int dimensions_count) {
+        return tflite::RuntimeShape(dimensions_count);
+      }
+    ))
+    .property("dims", &binding_utils::getRuntimeShapeDimensions, &binding_utils::setRuntimeShapeDimensions)
+    .property("size", &binding_utils::getRuntimeShapeSize, &binding_utils::setRuntimeShapeSize)
+    ;
+
+  class_<tflite::ConvParams>("ConvParams")
```
[the remark this answers is not in the document]
Comment (Wenzhao-Xiang) [reply src/nn/wasm/src/bind/src/binding.cpp]: I didn't notice the [`Value Objects`](https://kripken.github.io/emscripten-site/docs/porting/connecting_cpp_and_javascript/embind.html#value-types) before. Seems it really makes the code better! Thanks very much!

File: src/nn/wasm/src/bind/test/operations.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
+describe('Operations Test', function() {
```
[the remark this answers is not in the document]
Comment (Wenzhao-Xiang) [reply src/nn/wasm/src/bind/test/operations.js]: No,  it is out of date.
Since we can now test the ops directly by `npm test`, can I just remove it?

File: src/nn/wasm/src/bind/src/binding.cpp [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+using namespace emscripten;
+
+namespace binding_utils {
+  struct ReshapeParams {
+    uint32_t size_count;
+  };
+
+  // Factors set and get function
+  val getRuntimeShapeDimensions(const tflite::RuntimeShape& shape) {
+    emscripten::val js_dims = emscripten::val::array();
```
[the remark this answers is not in the document]
Comment (Wenzhao-Xiang) [reply src/nn/wasm/src/bind/src/binding.cpp]: Done!

File: src/nn/wasm/src/bind/test/operations.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (huningxin) [reply src/nn/wasm/src/bind/test/operations.js]: then please remove it, thanks!

File: src/nn/wasm/src/bind/src/binding.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (huningxin) [reply src/nn/wasm/src/bind/src/binding.cpp]: however, it seems every function returns `true`, any other cases?

[the remark this answers is not in the document]
Comment (Wenzhao-Xiang) [reply src/nn/wasm/src/bind/src/binding.cpp]: In fact, wasm will throw errors too, when calculating fail. I will remove the returns.

**Comment by Wenzhao-Xiang** [comment]
@huningxin  All done. Please have another look, thanks!

File: src/nn/wasm/src/bind/src/binding.cpp [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+  bool softmaxFloat32Wrapper(const SoftmaxParams op_params,
+                             const RuntimeShape& inputShape, 
+                             const intptr_t inputData, 
+                             const RuntimeShape& outputShape, 
+                             intptr_t outputData) {
+    optimized_ops::Softmax(op_params, inputShape, (const float*)inputData,
+                           outputShape, (float*)outputData);
+  }
+
+  bool reshapeGenericWrapper(const binding_utils::ReshapeParams op_params,
```
Comment (huningxin) [inline src/nn/wasm/src/bind/src/binding.cpp]: Why not compute the size by `inputShape`?

File: src/nn/wasm/src/bind/src/binding.cpp [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+  };
+
+  // Factors set and get function
+  val getRuntimeShapeDimensions(const RuntimeShape& shape) {
+    return emscripten::val(
+      emscripten::typed_memory_view(shape.DimensionsCount(),
+                                    shape.DimsData()));
+  }
+
+  void setRuntimeShapeDimensions(RuntimeShape& shape, val js_dims) {
```
Comment (huningxin) [inline src/nn/wasm/src/bind/src/binding.cpp]: I guess we don't need this method as `getRuntimeShapeDimensions` returns the `Uint32Array` of the `DimsData`. Could you test whether you can assign value to the `Uint32Array` directly?

Comment (huningxin) [reply src/nn/wasm/src/bind/src/binding.cpp]: Do we need resize the dims?

File: src/nn/wasm/src/bind/src/binding.cpp [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+    std::vector<int> tmp = vecFromJSArray<int32_t>(js_dims);
+    for (int i = 0; i < tmp.size(); i++) {
+      shape.SetDim(i, tmp.at(i));
+    }
+  }
+
+  val getRuntimeShapeSize(const RuntimeShape& shape) {
```
Comment (huningxin) [inline src/nn/wasm/src/bind/src/binding.cpp]: ditto. Perhaps we can use the `Uint32Array.length` directly.

File: src/nn/wasm/src/bind/src/binding.cpp [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+    emscripten::val js_val = val(shape.DimensionsCount());
+    return js_val;
+  }
+
+  void setRuntimeShapeSize(RuntimeShape& shape, int32_t size) {
```
Comment (huningxin) [inline src/nn/wasm/src/bind/src/binding.cpp]: Do we need resize the dims?

File: src/nn/wasm/src/bind/src/binding.cpp (hunk 7 above)
Comment (Wenzhao-Xiang) [reply src/nn/wasm/src/bind/src/binding.cpp]: I calculate it by `inputShape` in js, just to make all the init steps do in the js code. 
```js
if (utils.isTensor(input.type)) {
    size_count = utils.sizeOfTensorData(input.type, input.runtimeshape.dims);
} else {
    size_count = utils.sizeOfScalarData(input.type);
}
```
Move it to C++ code?

Comment (huningxin) [reply src/nn/wasm/src/bind/src/binding.cpp]: In this way, I suppose you don't need introduce `ReshapeParams`. It would make the binding code simpler, won't it?

**Comment by huningxin** [comment]
Thanks so much! Looks good to me.
